src/lsm/block_free_set.zig

```diff
@@ -1,37 +1,421 @@
+    try test_acquire_release(BlockFreeSet.shard_size);
+    try test_acquire_release(2 * BlockFreeSet.shard_size);
+    try test_acquire_release(63 * BlockFreeSet.shard_size);
+    try test_acquire_release(64 * BlockFreeSet.shard_size);
+    try test_acquire_release(65 * BlockFreeSet.shard_size);
+}
+
+fn test_acquire_release(blocks_count: usize) !void {
+    const expectEqual = std.testing.expectEqual;
+    // Acquire everything, then release, then acquire again.
+    var set = try BlockFreeSet.init(std.testing.allocator, blocks_count);
+    defer set.deinit(std.testing.allocator);
+
+    var empty = try BlockFreeSet.init(std.testing.allocator, blocks_count);
+    defer empty.deinit(std.testing.allocator);
+
+    var i: usize = 0;
+    while (i < blocks_count) : (i += 1) try expectEqual(@as(?u64, i + 1), set.acquire());
+    try expectEqual(@as(?u64, null), set.acquire());
+
+    i = 0;
+    while (i < blocks_count) : (i += 1) set.release(@as(u64, i + 1));
+    try expect_block_free_set_equal(empty, set);
+
+    i = 0;
+    while (i < blocks_count) : (i += 1) try expectEqual(@as(?u64, i + 1), set.acquire());
+    try expectEqual(@as(?u64, null), set.acquire());
+
+    // Exercise the RingBuffer-index sync by de/re-allocating the last bit of a shard.
+    set.release(i);
+    try expectEqual(@as(usize, 1), set.recent.count);
+    try expectEqual(@as(?u64, i), set.acquire());
+    try expectEqual(@as(usize, 0), set.recent.count);
+    try expectEqual(@as(?u64, null), set.acquire());
+}
+
+fn test_block_shards_count(expect_shards_count: usize, blocks_count: usize) !void {
+    var set = try BlockFreeSet.init(std.testing.allocator, blocks_count);
+    defer set.deinit(std.testing.allocator);
+
+    try std.testing.expectEqual(expect_shards_count, set.index.bit_length);
+}
+
+test "BlockFreeSet encode, decode, encode" {
+    const shard_size = BlockFreeSet.shard_size / @bitSizeOf(usize);
+    // Uniform.
+    try test_encode(&.{.{.fill = .uniform_ones, .words = shard_size}});
+    try test_encode(&.{.{.fill = .uniform_zeros, .words = shard_size}});
+    try test_encode(&.{.{.fill = .literal, .words = shard_size}});
+    try test_encode(&.{.{.fill = .uniform_ones, .words = std.math.maxInt(u16) + 1}});
+
+    // Mixed.
+    try test_encode(&.{
+        .{.fill = .uniform_ones, .words = shard_size / 4},
+        .{.fill = .uniform_zeros, .words = shard_size / 4},
+        .{.fill = .literal, .words = shard_size / 4},
+        .{.fill = .uniform_ones, .words = shard_size / 4},
+    });
+
+    // Random.
+    var seed: u64 = undefined;
+    try std.os.getrandom(mem.asBytes(&seed));
+    var prng = std.rand.DefaultPrng.init(seed);
+
+    const fills = [_]TestPatternFill{.uniform_ones, .uniform_zeros, .literal};
+    var t: usize = 0;
+    while (t < 10) : (t += 1) {
+        var patterns = std.ArrayList(TestPattern).init(std.testing.allocator);
+        defer patterns.deinit();
+        var i: usize = 0;
```

Comment:
```suggestion

        var i: usize = 0;
```

Just to make the defer above stand out.

Reply:
We want to keep resource allocation/deallocation grouped so that it's easy to see the defer is not missing.